### PR TITLE
Add mode to `int_as_pointer`

### PR DESCRIPTION
**Pull request by riaqn**
This PR adds a mode parameter to the `int_as_pointer` primitive, so that when you write:
```
external follow : int -> ('a [@local_opt]) = "%int_as_pointer"
```
that parameter will reflect the mode of `'a`.  If the mode is `local`, `follow` is understood to be allocating in the current region, and the current region must be preserved by optimizations. This PR changes the behaviours around `int_as_pointer` to reflect that. The behaviours of `int_as_pointer` itself is not changed. 

Request review from @mshinwell or @lpw25

### CODE REVIEW

File: middle_end/flambda2/terms/flambda_primitive.ml
Comment (lpw25) [inline middle_end/flambda2/terms/flambda_primitive.ml:1001]: You should lift this out to it's own function and also use it for e.g. `Box_number`.

File: ocaml/testsuite/tests/typing-local/regions.ml
Comment (lpw25) [inline ocaml/testsuite/tests/typing-local/regions.ml:21]: Slightly better to use an informative name here. How about `allocate_in_parent_region` for this one and `leak` for `foo_bad`?

File: ocaml/testsuite/tests/typing-local/regions.ml
Comment (lpw25) [inline ocaml/testsuite/tests/typing-local/regions.ml:33]: I think using `0` here means getting a pointer that might explode in no-naked-pointers mode. You're probably going to need to write a C stub that creates a genuine external pointer.

File: ocaml/testsuite/tests/typing-local/regions.ml
Comment (lpw25) [inline ocaml/testsuite/tests/typing-local/regions.ml:46]: I think you should create a `check_not_empty` that prints `OK` if the thing leaks and something like `Failed` if it doesn't. Otherwise the reference makes it look like the test is broken. You should be able to factor out the common part of `check_empty` and `check_not_empty` so that there isn't much new code.

**Review by lpw25 — APPROVED** [review]
There's a bug that needs fixing and some small suggestions. Good to merge once they are addressed.

File: ocaml/testsuite/tests/typing-local/regions.ml
Comment (riaqn) [reply ocaml/testsuite/tests/typing-local/regions.ml]: I wrote a C stub and it seems to be working. However, I didn't use the ocaml5-proof approach that you mentioned because I forgot.

**Comment by riaqn** [comment]
@lpw25 it seems that this PR touches part of the code base and requires more approval. Can you override this and merge?

**Review by lpw25 — APPROVED** [review]

**Pull request merged**
